src/chat/chatReducers.js

```diff
@@ -149,6 +152,63 @@ const eventUpdateMessage = (
     last_edit_timestamp: action.edit_timestamp,
   }));
 
+const eventUpdateMessageFlags = (
+  state: MessagesState,
+  action: EventUpdateMessageFlagsAction,
+): MessagesState => {
+  /**
+    In order to add new messages to the 'is:starred' narrow, we cache newly
+    starred messages.
+
+    This local variable will no longer be necessary when the messages state is
+    refactored into a map of message_id -> message, as narrows will then be
+    represented as lists of ids, and referencing action.messages will suffice.
+  */
+  const messageCache: { [id: number]: Message } = {};
+
+  // Update the flags fields
+  const result = Object.keys(state).reduce((updatedState, narrow) => {
+    updatedState[narrow] = state[narrow].map(message => {
+      const updatedMessage = Object.assign({}, message);
+
+      if (action.messages.includes(message.id)) {
+        if (action.operation === 'add') {
+          updatedMessage.flags = [...message.flags, action.flag];
+          if (action.flag === 'starred') {
+            // local side effect: cache newly starred messages
+            messageCache[message.id] = updatedMessage;
+          }
+        } else {
+          // action.operation === 'remove'
+          updatedMessage.flags = message.flags.filter(flag => flag !== action.flag);
+        }
+      }
+
+      return updatedMessage;
+    });
+
+    return updatedState;
+  }, {});
+
+  // Update the 'is:starred' narrow
```

Comment:
The `is:mentioned` narrow needs the same logic, I think.

Hrm, but the webapp doesn't seem to do anything with flags other than `starred` and `read`. In fact the code is kind of beautifully simple, just a few lines:
https://github.com/zulip/zulip/blob/32438cde9/static/js/server_events_dispatch.js#L402

... wait. It doesn't actually work. Basically it has a type error. :-P And, empirically, if I'm looking at some messages in the webapp and I go and star or unstar some of them from mobile, or from another tab on web -- nothing happens. I guess I'll go file that bug in zulip/zulip.

There's another lesson here, though: the reason nobody noticed that was broken until now is that the webapp has *local echo* for starring a message.

We should do local echo too, independent of this fix; on a slow network, or if you're just on a different continent from your server, you don't want to wait 200ms to see the star appear.

Reply:
Also, none of this kind of stuff is documented, but I think the server may not send this event (relying on `update_message` instead?) for a normal `mentioned` change. So let's not worry about that one for now.

Reply:
I've opened #2852.

I've never heard of the term *local echo* before - a quick [wikipedia search](https://en.wikipedia.org/wiki/Echo_(computing)) provides context:
> ... **local echo**, where the sending device itself displays the sent data

If I'm understanding that correctly, I think what you're referring to here is identical in implementation to an '[optimistic update](https://stackoverflow.com/questions/33009657/what-is-optimistic-updates-in-front-end-development)'.

Reply:
Ah, indeed!

"Optimistic update" isn't a term I know / that I see used in the communities I know. But yes, that's exactly what I mean.

As that Wikipedia article suggests, the term "local echo" comes from talking about interactions where the user is typing some text, and that text gets "echoed" locally without waiting for the server to echo it back. This turns out to make a big difference in the pleasantness of the interaction, if the server is more than a few ms away; it's one of the major features [mosh](https://mosh.org/) is proud of in contrast with `ssh`. :)

We use it a lot in the Zulip project in describing how when the user sends a message, we display the message immediately rather than wait for it to come back from the server. Search for "locally echo" here:
https://zulip.readthedocs.io/en/latest/subsystems/markdown.html